### PR TITLE
Make it possible to include custom columns from TestActions in case, class and package tables

### PR DESCRIPTION
**Pull request by jonaslind**
Add optional jelly hooks that allow TestActions to contribute custom columns to tables that list CaseResults, ClassResults and PackageResults.

More details in the corresponding enhancement issue:
Fixes https://github.com/jenkinsci/junit-plugin/issues/494

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by jonaslind** [comment]
Desired reviewer: @timja

**Comment by timja** [comment]
can you check the build is green before requesting my review please

**Comment by jonesbusy** [comment]
Looks also the master branch is broken : https://ci.jenkins.io/job/Plugins/job/junit-plugin/job/master/321/

**Comment by jonaslind** [comment]
@timja @jonesbusy thanks for comments!

Yes, the reason the build fails is RequireUpperBoundDeps, and this issue comes from the master branch. It's not introduced by my pull request. For example, I saw the same failure in @timja 's pull request #501 that was merged yesterday:

https://ci.jenkins.io/job/Plugins/job/junit-plugin/view/change-requests/job/PR-501/1/

It looks like the issue has been introduced by dependabot since c789200cf343cef9d806e394f050eac104c700c0. At the same time, it looks like @timja is replacing dependabot with renovate?

I'm happy to try and tackle the RequireUpperBoundDeps failure if no-one gets to it before me. Am I expected to fix that in this pull request, or shall it be done separately in a different pull request?

**Comment by timja** [comment]
Weird given CI passed on that build.

I've switched it to renovate but they are basically the same tools with a slightly different feature set.

If you're happy to do it then a separate PR would be great. Otherwise I'll get to it when I have a chance

**Comment by jonesbusy** [comment]
@jonaslind I'm not maintainer of this plugin, but same things happen to me on the skip-notifications-trait-plugin 

https://github.com/jenkinsci/skip-notifications-trait-plugin/pull/62/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R59

**Comment by jonaslind** [comment]
@timja , do you think you'll be able to review this pull request?

**Comment by timja** [comment]
At some point yes but I don't have time right now, I've requested review from all the maintainers though

File: src/main/java/hudson/tasks/junit/TestAction.java
Comment (timja) [inline src/main/java/hudson/tasks/junit/TestAction.java:37]:
```suggestion
 * <li>classresult-tableheader.jelly: allows additional table headers to be shown in the test class's table of test methods</li>
 * <li>classresult-tablerow.jelly: allows additional table cells to be shown in the test class's table of test methods</li>
```

**Review by timja — APPROVED** [review]
Seems fine, is there a reason you're just doing this for class results and not other ones?

File: src/main/java/hudson/tasks/junit/TestAction.java
Comment (jonaslind) [reply src/main/java/hudson/tasks/junit/TestAction.java:37]: Thank you for the grammar fix, will include in next commit!

**Comment by timja** [comment]
I think they are as you click through test rests from the class level to the individual test level.

Its often hard to tell where they are used, I generally put some gibberish in each file and click around till I trace them back

**Comment by jonaslind** [comment]
@timja, here's a new commit with custom columns added to `MetaTabulatedResult`. Turns out that `History` is not a `TestObject`, so adding custom columns to history pages would involve more refactoring. I'm not sure that is motivated unless there is a clear use case for it. So I've skipped that for now.

Please have a look when you have the time. Thanks!

File: src/main/java/hudson/tasks/test/TabulatedResult.java
Comment (timja) [inline src/main/java/hudson/tasks/test/TabulatedResult.java:125]: please add javadoc

**Review by timja — COMMENTED** [review]
looks good, just one docs change missing

**Review by timja — APPROVED** [review]
Thanks!

**Comment by jonaslind** [comment]
Thank you very much @timja for all your help and input!

**Pull request merged**
